### PR TITLE
rewrite config from scratch, add --dry-run option that skips ftp upload

### PR DESCRIPTION
**Pull request by Fogapod**

### CODE REVIEW

File: usautobuild/config.py [hunk 1]
```diff
@@ -1,127 +1,171 @@
...
+        self,
+        default: Any = _UNSET,
+        *,
+        type_: Optional[Any] = None,
+        arg: Optional[str] = None,
+        config: Optional[str] = None,
+        env: Optional[str] = None,
+    ):
+        self.default = default
+        self.type_ = type_
```
Comment (corp-0) [inline usautobuild/config.py:37]: why is the underscore to the right?

Comment (Fogapod) [reply usautobuild/config.py:37]: because it otherwise shadows builtin `type` function

File: usautobuild/config.py [hunk 2]
```diff
@@ -1,127 +1,171 @@
...
+    abort_on_build_fail = Var(True)
+    allow_no_changes = Var(True)
+
+    build_number = Var(0)
+
+    output_dir = Var(Path.cwd() / "builds")
+    license_file = Var(Path.cwd() / "UnityLicense.ulf")
+    project_path = Var(Path())
+
+    def __init__(self, args: dict[str, Any]):
```
Comment (corp-0) [inline usautobuild/config.py:134]: this init method is chonky, isn't it?

Comment (Fogapod) [reply usautobuild/config.py:134]: It's not super complex. I agree that it's a good idea to put code somewhere else but i had no time today

File: usautobuild/config.py [hunk 3]
```diff
@@ -1,127 +1,171 @@
...
+                else:
+                    value = True
+            elif type_ is list:
+                value = value.split(",")
+
+        if (config := self.config) is None:
+            config = name
+
+        value = cfg.get(config, value)
+        value = args.get(config, value)
```
Comment (Fogapod) [inline usautobuild/config.py:70]: there is a bug: should be `get(arg, value)`, same as with config but it isn't used anywhere now anyway

Comment (Fogapod) [reply usautobuild/config.py]: fixed

File: usautobuild/config.py (hunk 2 above)
Comment (Fogapod) [reply usautobuild/config.py:162]: fixed

**Review by corp-0 — APPROVED** [review]

**Pull request merged**